README.md
# flutter_wallet

[![pub package](https://img.shields.io/pub/v/flutter_wallet.svg)](https://pub.dartlang.org/packages/flutter_wallet)

Flutter plugin to add pkpass to iOS wallet (Passbook)

## Usage

To use, you must at least specify one arguments **pkpass**.
* The argument **pkpass** is List<int> type.
  
### Example
  
```dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter_wallet/flutter_wallet.dart';
import 'package:dio/dio.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  Future<void> getPass() async {
    String jsonParameters = '{';
    jsonParameters += '"serialNumber" : "123456789",';
    jsonParameters += '"relevantDate" : "2019-07-20T15:30:00-06:00",';
    jsonParameters += '"latitude" : 19.687,';
    jsonParameters += '"longitude" : -101.151,';
    jsonParameters += '"relevantText" : "Wellcome';
    jsonParameters += '"message" : "QQ2475P",';
    jsonParameters += '"date" : "20 Jul.",';
    jsonParameters += '"time" : "03:30 PM",';
    jsonParameters += '"movie" : "Movie Name",';
    jsonParameters += '"cinema" : "Cinema Name",';
    jsonParameters += '"seats" : "A1",';
    jsonParameters += '"room" : "10",';
    jsonParameters += '"poster" : "https://domain.com/images/image.jpg"';
    jsonParameters += '}';

    Response response;
    Dio dio = new Dio();

    response = await dio.post("http://domain.com/getPKPass.php", 
                          data: jsonParameters, 
                          options: Options(responseType: ResponseType.bytes));

    if (response.data != null) {
      try {
        var result = await FlutterWallet.addPass(pkpass: response.data);
        print(result);
      } catch (e) {
        print(e.message);
      }
    } else {
       print("Data error");
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Passbook example app'),
        ),
        body: Center(
          child: RaisedButton(child: Text("Get Passbook"), onPressed: () {
            getPass();
          },),
        ),
      ),
    );
  }
}
```

## Notes

Don't forget enable Wallet capability in your Xcode project:

![](https://github.com/vico-aguado/flutter_wallet/blob/master/capability.png)

And Wallet service in your application service 
(Certificates, Identifiers & Profiles / Identifiers / App IDs / 'Your app id'):

![](https://github.com/vico-aguado/flutter_wallet/blob/master/appService.png)

## TODO:
* Add Android compatibility.
* Add more documentation.

------------------------------------------------------------------------------

## Getting Started

This project is a starting point for a Flutter
[plug-in package](https://flutter.io/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter, view our 
[online documentation](https://flutter.io/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
# Changelog

## 0.0.1 - 2019-04-28

* initial release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_wallet_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_wallet_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_wallet/flutter_wallet.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_wallet');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

/*
  test('getPlatformVersion', () async {
    expect(await FlutterWallet.platformVersion, '42');
  });*/
}


lib/flutter_wallet.dart
/**
 * 
Flutter Wallet

Copyright (c) 2019 the Flutter Wallet authors.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.
 * Neither the name of the flutter wallet authors, nor the
   names of its contributors may be used to endorse or promote products
   derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 */
import 'dart:async';
import 'package:flutter/services.dart';
import 'package:meta/meta.dart';

class FlutterWallet {
  static const MethodChannel _channel =
      const MethodChannel('com.vico-aguado.flutter/wallet');

  /// Returns bool value if iOS can covert one pkpass file
  ///
  /// - parameters:
  ///    - pkpass: List with <int> values from pkpass file 
  static Future<bool> addPass({@required List<int> pkpass}) async {
    final bool result = await _channel
        .invokeMethod('addWalletPass', <String, dynamic>{'pkpass': pkpass});
    return result;
  }
}


ios/Classes/FlutterWalletPlugin.m
#import "FlutterWalletPlugin.h"
#import <PassKit/PassKit.h>

@implementation FlutterWalletPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"com.vico-aguado.flutter/wallet"
            binaryMessenger:[registrar messenger]];
  FlutterWalletPlugin* instance = [[FlutterWalletPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"addWalletPass" isEqualToString:call.method]) {
      
      if (call.arguments != nil) {
          if ([call.arguments[@"pkpass"] isEqual:[NSNull null]]) {
              result([FlutterError errorWithCode:@"WITHOUT_PARAMETERS" message:@"Don't have 'pkpass' parameter" details:@"You need add 'pkpass' parameter"]);
          } else {
              if ([call.arguments[@"pkpass"] isKindOfClass:[NSArray class]]) {
                  NSArray *pkpassArray = call.arguments[@"pkpass"];
                  if ([pkpassArray count] > 0) {
                      NSMutableData *data = [[NSMutableData alloc] initWithCapacity: [pkpassArray count]];
                      for( NSNumber *number in pkpassArray) {
                          char byte = [number charValue];
                          [data appendBytes: &byte length: 1];
                      }
                      
                      NSError* errorPass = nil;
                      PKPass *newPass = [[PKPass alloc] initWithData:data
                                                               error:&errorPass];
                      
                      if (errorPass!=nil) {
                          result([FlutterError errorWithCode:@"PKPASS_ERROR" message:[errorPass localizedDescription] details:nil]);
                      }else{
                          UIViewController *rootController = [UIApplication sharedApplication].delegate.window.rootViewController;
                          
                          if (rootController != nil) {
                              PKAddPassesViewController *addController = [[PKAddPassesViewController alloc] initWithPass:newPass];
                              [rootController presentViewController:addController animated:YES completion:nil];
                          } else {
                              result([FlutterError errorWithCode:@"ROOTCONTROLLER_INVALID" message:@"rootViewController of the app is invalid" details:nil]);
                          }
                          result([NSNumber numberWithBool:YES]);
                      }
                  } else {
                      result([FlutterError errorWithCode:@"PARAMETERS_INVALID" message:@"Your 'pkpass' parameter is invalid" details:nil]);
                  }
              } else {
                  result([FlutterError errorWithCode:@"PARAMETERS_INVALID" message:@"Your 'pkpass' parameter is invalid" details:nil]);
              }
          }
      }else{
          result([FlutterError errorWithCode:@"WITHOUT_PARAMETERS" message:@"Don't have 'pkpass' parameter" details:@"You need add 'pkpass' parameter"]);
      }
      
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end


android/src/main/java/com/vicoaguado/flutter_wallet/FlutterWalletPlugin.java
package com.vicoaguado.flutter_wallet;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterWalletPlugin */
public class FlutterWalletPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_wallet");
    channel.setMethodCallHandler(new FlutterWalletPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter_wallet/flutter_wallet.dart';
import 'package:dio/dio.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  Future<void> getPass() async {
    String jsonParameters = '{';
    jsonParameters += '"serialNumber" : "123456789",';
    jsonParameters += '"relevantDate" : "2019-07-20T15:30:00-06:00",';
    jsonParameters += '"latitude" : 19.687,';
    jsonParameters += '"longitude" : -101.151,';
    jsonParameters += '"relevantText" : "Wellcome';
    jsonParameters += '"message" : "QQ2475P",';
    jsonParameters += '"date" : "20 Jul.",';
    jsonParameters += '"time" : "03:30 PM",';
    jsonParameters += '"movie" : "Movie Name",';
    jsonParameters += '"cinema" : "Cinema Name",';
    jsonParameters += '"seats" : "A1",';
    jsonParameters += '"room" : "10",';
    jsonParameters += '"poster" : "https://domain.com/images/image.jpg"';
    jsonParameters += '}';

    Response response;
    Dio dio = new Dio();

    response = await dio.post("http://domain.com/getPKPass.php", data: jsonParameters, options: Options(responseType: ResponseType.bytes));

    if (response.data != null) {
      try {
        var result = await FlutterWallet.addPass(pkpass: response.data);
        print(result);
      } catch (e) {
        print(e.message);
      }
    } else {

    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Passbook example app'),
        ),
        body: Center(
          child: RaisedButton(child: Text("Get Passbook"), onPressed: () {
            getPass();
          },),
        ),
      ),
    );
  }
}
